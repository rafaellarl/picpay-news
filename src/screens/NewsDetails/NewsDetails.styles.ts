import styled from 'styled-components/native';

export const Wrapper = styled.ScrollView`
    flex: 1;
    background-color: #111111;
`;

export const Container = styled.View`
    flex: 1;
    padding: 16px;
`;

export const Image = styled.Image`
    width: 100%;
    height: 200px;
`;

export const Title = styled.Text.attrs({
    numberOfLines: 3,
    ellipsizeMode: 'tail'
})`
    font-size: 20px;
    line-height: 26px;
    font-weight: 400;
    color: #FFFFFF;
    text-align: center;
`;

export const Description = styled.Text`
    font-size: 16px;
    line-height: 20px;
    font-weight: 400;
    color: #FFFFFF;
    padding-top: 32;
`;

export const DescriptionBold = styled(Description)`
    font-weight: 700;
`;