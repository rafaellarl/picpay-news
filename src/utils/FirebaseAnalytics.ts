import Analytics from '@react-native-firebase/analytics';

/*
    Melhorias: Fazer o uso de tipagem mais restrita (ex: union types ou enums) para limitar os valores possíveis de `screenName` e `flow`.
    Isso ajuda a evitar erros de digitação e torna o código mais seguro e fácil de manter.
*/
interface IScreenView {
  screenName: string;
  flow: string;
}

interface ISelectContent extends IScreenView {
  contentType: string;
}

interface IException extends IScreenView {
  description: string;
  fatal?: boolean;
}

interface ISuccess extends IScreenView {
  description: string;
}

interface ISelectAds extends IScreenView {
  idAds: string;
}

/*
    Decisão: Optei por manter funções separadas para cada tipo de evento, 
    mesmo sendo possível consolidar em uma única função com `eventName` como parâmetro.
    Essa abordagem reduz o risco de envio incorreto de eventos e melhora a legibilidade 
    e manutenção do código em projetos maiores.
*/
export default class FirebaseAnalytics {
  static saveScreenView = async ({screenName, flow}: IScreenView) => {
    const parameters = {
      flow,
      screen_name: screenName,
    };

    Analytics().logEvent('screen_view', parameters);
  };

  static saveSelectContent = async ({
    screenName,
    flow,
    contentType,
  }: ISelectContent) => {
    const parameters = {
      flow,
      screen_name: screenName,
      content_type: contentType,
    };

    Analytics().logEvent('select_content', parameters);
  };

  static saveEventException = async ({
    flow,
    screenName,
    fatal = false,
    description = '',
  }: IException) => {
    const parameters = {
      flow,
      fatal: fatal,
      screen_name: screenName,
      description: description,
    };
    Analytics().logEvent('exception', parameters);
  };

  static saveEventSuccess = async ({
    flow,
    screenName,
    description,
  }: ISuccess) => {
    const parameters = {
      flow,
      description,
      screen_name: screenName,
    };
    Analytics().logEvent('success', parameters);
  };

  static saveEventSelectAds = async ({flow, idAds, screenName}: ISelectAds) => {
    const parameters = {
      flow,
      id_ads: idAds,
      screen_name: screenName,
    };
    Analytics().logEvent('select_content', parameters);
  };
}
