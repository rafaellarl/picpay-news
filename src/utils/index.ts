export {default as FirebaseAnalytics} from './analytics/FirebaseAnalytics';
export {default as getAdUnitId} from './getAdUnitId/getAdUnitId';
export {default as formatDateToShortPtBr} from './formatDateToShortPtBr/formatDateToShortPtBr';
