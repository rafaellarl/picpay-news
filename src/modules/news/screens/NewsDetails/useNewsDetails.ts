import {Platform} from 'react-native';
import {useEffect, useRef} from 'react';
import {useRoute} from '@react-navigation/native';
import {BannerAd, useForeground} from 'react-native-google-mobile-ads';

import {ANALYTICS_SCREENS, FLOW} from '../../constants/analytics';
import {NewsDetailsScreenRouteProps} from '../../../../routes/types';
import {FirebaseAnalytics, getAdUnitId} from '../../../../utils';

const useNewsDetails = () => {
  const {
    params: {image, title},
  } = useRoute<NewsDetailsScreenRouteProps>();
  const adUnitId = getAdUnitId('ADAPTIVE_BANNER');
  const bannerRef = useRef<BannerAd>(null);

  useForeground(() => {
    Platform.OS === 'ios' && bannerRef.current?.load();
  });

  useEffect(() => {
    FirebaseAnalytics.saveScreenView({
      flow: FLOW,
      screenName: ANALYTICS_SCREENS.newsDetails.name,
    });
  }, []);

  return {
    image,
    title,
    adUnitId,
    bannerRef,
  };
};

export default useNewsDetails;
