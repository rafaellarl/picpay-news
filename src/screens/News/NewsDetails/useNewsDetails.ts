import {useEffect, useRef} from 'react';
import {Platform} from 'react-native';
import {useRoute} from '@react-navigation/native';
import {BannerAd, TestIds, useForeground} from 'react-native-google-mobile-ads';

import {NewsDetailsScreenRouteProp} from '../../../routes/types/newsTypes';
import FirebaseAnalytics from '../../../utils/FirebaseAnalytics';

const useNewsDetails = () => {
  const {
    params: {image, title},
  } = useRoute<NewsDetailsScreenRouteProp>();
  // TOODO: Melhorar essa parte deixar mais seguro
  const adUnitIdTest =
    Platform.OS === 'ios'
      ? 'ca-app-pub-4128753647068732~2668177669'
      : 'ca-app-pub-4128753647068732~6719103316';
  const adUnitId = __DEV__ ? TestIds.ADAPTIVE_BANNER : adUnitIdTest;
  const bannerRef = useRef<BannerAd>(null);

  useForeground(() => {
    Platform.OS === 'ios' && bannerRef.current?.load();
  });

  useEffect(() => {
    FirebaseAnalytics.saveScreenView({
      flow: 'new',
      screenName: 'new-details',
    });
  }, []);

  return {
    image,
    title,
    adUnitId,
    bannerRef,
  };
};

export default useNewsDetails;
