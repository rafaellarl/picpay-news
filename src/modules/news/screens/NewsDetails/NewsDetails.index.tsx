import React from 'react';
import {BannerAd, BannerAdSize} from 'react-native-google-mobile-ads';

import {labels} from '../../labels';
import {TEST_IDS_SCREENS} from '../../constants';

import {
  Image,
  Title,
  Wrapper,
  Container,
  Description,
  DescriptionBold,
} from './NewsDetails.styles';
import useNewsDetails from './useNewsDetails';

/*
  Sugestão: Em uma api bem formatada podemos passar o body da noticía por rota também, neste caso como é uma api free o body ficou limitado.
  Por esse motivo coloquei um texto gerado por IA falando sobre programar nativo e em frameworks multiplataformas.

  Colocando um Banner de anúncio para estudo da lib e apresentar outros formas de apresentar anúncios.
*/
const NewsDetails = () => {
  const {image, title, adUnitId, bannerRef} = useNewsDetails();

  return (
    <Wrapper>
      <Image src={image} testID={TEST_IDS_SCREENS.newsDetails.imageNews} />
      <Container>
        <Title>{title}</Title>
        <DescriptionBold>{labels.newsDetails.subTitle}</DescriptionBold>
        <Description>{labels.newsDetails.detailsPartI}</Description>
        <Description>{labels.newsDetails.detailsPartII}</Description>
        <Description>{labels.newsDetails.detailsPartIII}</Description>
      </Container>
      <BannerAd
        ref={bannerRef}
        unitId={adUnitId}
        size={BannerAdSize.ANCHORED_ADAPTIVE_BANNER}
      />
    </Wrapper>
  );
};

export default NewsDetails;
