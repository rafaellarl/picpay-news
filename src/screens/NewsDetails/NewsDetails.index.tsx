import React from "react";
import { BannerAd, BannerAdSize } from "react-native-google-mobile-ads";

import { Wrapper, Image, Container, Title, Description, DescriptionBold } from './NewsDetails.styles';
import useNewsDetails from "./useDetails";

const NewsDetails = () => {
    const {
        image,
        title,
        adUnitId,
        bannerRef,
    } = useNewsDetails()

    return (
        <Wrapper>
            <Image src={image} />
            <Container>
                <Title>{title}</Title>
                <DescriptionBold>{'Texto gerado por IA'}</DescriptionBold>
                <Description>
                    {'Plataformas nativas são aquelas desenvolvidas especificamente para cada sistema operacional, como Android e iOS. Essa abordagem permite um desempenho otimizado e acesso total aos recursos do dispositivo, embora exija a manutenção de códigos separados para cada plataforma. Por outro lado, as plataformas híbridas utilizam um único código-base para funcionar em diferentes sistemas, facilitando a manutenção e reduzindo os custos de desenvolvimento. Frameworks como React Native e Flutter são exemplos populares que têm ganhado espaço por suas capacidades e eficiência. O React Native permite a criação de interfaces com uma experiência próxima à nativa, utilizando JavaScript e uma comunidade ativa, enquanto o Flutter, com base em Dart, oferece alta performance e uma rica personalização visual por meio de seus widgets.'}
                </Description>
                <Description>
                    {'Para desenvolvedores que buscam agilidade e versatilidade, trabalhar com frameworks híbridos pode ser uma ótima escolha, principalmente se o projeto não exigir funcionalidades extremamente específicas de cada sistema operacional. A escolha entre uma plataforma nativa e uma híbrida dependerá do perfil do aplicativo, dos requisitos de performance e da experiência desejada para o usuário final. Assim, é importante avaliar cuidadosamente as necessidades do projeto e a disponibilidade de recursos para a manutenção do código.'}
                </Description>
                <Description>
                    {'Em suma, a decisão entre programação nativa e híbrida envolve ponderar performance, custo e facilidade de manutenção. Compreender as vantagens de cada abordagem, assim como dominar ferramentas como React Native e Flutter, permite ao desenvolvedor tomar decisões mais estratégicas. Experimente baixar um ambiente de desenvolvimento móvel, como o Android Studio ou o Visual Studio Code com plugins específicos para mobile, e teste na prática qual abordagem melhor se adapta aos seus projetos. O universo do desenvolvimento mobile está repleto de oportunidades para inovar e transformar ideias em aplicativos eficientes e impactantes.'}
                </Description>
            </Container>
            <BannerAd
                ref={bannerRef}
                unitId={adUnitId}
                size={BannerAdSize.ANCHORED_ADAPTIVE_BANNER}
            />
        </Wrapper>
    )
};

export default NewsDetails;
