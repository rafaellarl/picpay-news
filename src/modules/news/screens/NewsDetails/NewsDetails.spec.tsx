import {useRoute} from '@react-navigation/native';
import {render} from '@testing-library/react-native';

import NewsDetails from './NewsDetails.index';
import {FirebaseAnalytics} from '../../../../utils';

jest.mock('@react-navigation/native', () => ({
  useRoute: jest.fn(),
}));

jest.mock('@react-native-firebase/analytics', () => ({
  default: jest.fn(() => ({
    logEvent: jest.fn(),
  })),
}));

jest.mock('../../../../utils/analytics/FirebaseAnalytics', () => ({
  saveScreenView: jest.fn(),
}));

const mockedUseRoute = useRoute as jest.Mock;
const mockedSaveScreenView = FirebaseAnalytics.saveScreenView as jest.Mock;

const renderScreen = () => render(<NewsDetails />);

describe('NewsDetails', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    mockedUseRoute.mockReturnValue({
      params: {
        image:
          'https://static.escolakids.uol.com.br/2025/06/ilustracao-de-um-menino-com-a-mao-na-orelha-captando-ondas-sonoras-ou-de-som.jpg',
        title: 'Som: o que é, como é produzido, curiosidades - Escola Kids',
      },
    });
  });

  it('should match snapshot on initial render', () => {
    const {toJSON} = renderScreen();
    expect(toJSON()).toMatchSnapshot();
  });

  it('Should render news details', () => {
    const {getByText, getByTestId} = renderScreen();

    getByTestId('test-id-news-image');
    getByText('Som: o que é, como é produzido, curiosidades - Escola Kids');
    getByText('Texto gerado por IA');
    getByText(
      'Plataformas nativas são aquelas desenvolvidas especificamente para cada sistema operacional, como Android e iOS. Essa abordagem permite um desempenho otimizado e acesso total aos recursos do dispositivo, embora exija a manutenção de códigos separados para cada plataforma. Por outro lado, as plataformas híbridas utilizam um único código-base para funcionar em diferentes sistemas, facilitando a manutenção e reduzindo os custos de desenvolvimento. Frameworks como React Native e Flutter são exemplos populares que têm ganhado espaço por suas capacidades e eficiência. O React Native permite a criação de interfaces com uma experiência próxima à nativa, utilizando JavaScript e uma comunidade ativa, enquanto o Flutter, com base em Dart, oferece alta performance e uma rica personalização visual por meio de seus widgets.',
    );
    getByText(
      'Para desenvolvedores que buscam agilidade e versatilidade, trabalhar com frameworks híbridos pode ser uma ótima escolha, principalmente se o projeto não exigir funcionalidades extremamente específicas de cada sistema operacional. A escolha entre uma plataforma nativa e uma híbrida dependerá do perfil do aplicativo, dos requisitos de performance e da experiência desejada para o usuário final. Assim, é importante avaliar cuidadosamente as necessidades do projeto e a disponibilidade de recursos para a manutenção do código.',
    );
    getByText(
      'Em suma, a decisão entre programação nativa e híbrida envolve ponderar performance, custo e facilidade de manutenção. Compreender as vantagens de cada abordagem, assim como dominar ferramentas como React Native e Flutter, permite ao desenvolvedor tomar decisões mais estratégicas. Experimente baixar um ambiente de desenvolvimento móvel, como o Android Studio ou o Visual Studio Code com plugins específicos para mobile, e teste na prática qual abordagem melhor se adapta aos seus projetos. O universo do desenvolvimento mobile está repleto de oportunidades para inovar e transformar ideias em aplicativos eficientes e impactantes.',
    );
  });

  it('Should register event screen view', () => {
    renderScreen();
    expect(mockedSaveScreenView).toHaveBeenCalledTimes(1);
    expect(mockedSaveScreenView).toHaveBeenCalledWith({
      flow: 'news',
      screenName: 'news_details',
    });
  });
});
