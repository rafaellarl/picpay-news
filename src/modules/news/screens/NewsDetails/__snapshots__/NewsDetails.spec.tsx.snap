// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`NewsDetails should match snapshot on initial render 1`] = `
<RCTScrollView
  style={
    {
      "backgroundColor": "#111111",
      "flexBasis": 0,
      "flexGrow": 1,
      "flexShrink": 1,
    }
  }
>
  <View>
    <Image
      src="https://static.escolakids.uol.com.br/2025/06/ilustracao-de-um-menino-com-a-mao-na-orelha-captando-ondas-sonoras-ou-de-som.jpg"
      style={
        {
          "height": 200,
          "width": "100%",
        }
      }
      testID="test-id-news-image"
    />
    <View
      style={
        {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "paddingBottom": 16,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
        }
      }
    >
      <Text
        ellipsizeMode="tail"
        numberOfLines={3}
        style={
          {
            "color": "#ffffff",
            "fontSize": 20,
            "fontWeight": "400",
            "lineHeight": 26,
            "textAlign": "center",
          }
        }
      >
        Som: o que é, como é produzido, curiosidades - Escola Kids
      </Text>
      <Text
        style={
          {
            "color": "#ffffff",
            "fontSize": 16,
            "fontWeight": "700",
            "lineHeight": 20,
            "paddingTop": 32,
          }
        }
      >
        Texto gerado por IA
      </Text>
      <Text
        style={
          {
            "color": "#ffffff",
            "fontSize": 16,
            "fontWeight": "400",
            "lineHeight": 20,
            "paddingTop": 32,
          }
        }
      >
        Plataformas nativas são aquelas desenvolvidas especificamente para cada sistema operacional, como Android e iOS. Essa abordagem permite um desempenho otimizado e acesso total aos recursos do dispositivo, embora exija a manutenção de códigos separados para cada plataforma. Por outro lado, as plataformas híbridas utilizam um único código-base para funcionar em diferentes sistemas, facilitando a manutenção e reduzindo os custos de desenvolvimento. Frameworks como React Native e Flutter são exemplos populares que têm ganhado espaço por suas capacidades e eficiência. O React Native permite a criação de interfaces com uma experiência próxima à nativa, utilizando JavaScript e uma comunidade ativa, enquanto o Flutter, com base em Dart, oferece alta performance e uma rica personalização visual por meio de seus widgets.
      </Text>
      <Text
        style={
          {
            "color": "#ffffff",
            "fontSize": 16,
            "fontWeight": "400",
            "lineHeight": 20,
            "paddingTop": 32,
          }
        }
      >
        Para desenvolvedores que buscam agilidade e versatilidade, trabalhar com frameworks híbridos pode ser uma ótima escolha, principalmente se o projeto não exigir funcionalidades extremamente específicas de cada sistema operacional. A escolha entre uma plataforma nativa e uma híbrida dependerá do perfil do aplicativo, dos requisitos de performance e da experiência desejada para o usuário final. Assim, é importante avaliar cuidadosamente as necessidades do projeto e a disponibilidade de recursos para a manutenção do código.
      </Text>
      <Text
        style={
          {
            "color": "#ffffff",
            "fontSize": 16,
            "fontWeight": "400",
            "lineHeight": 20,
            "paddingTop": 32,
          }
        }
      >
        Em suma, a decisão entre programação nativa e híbrida envolve ponderar performance, custo e facilidade de manutenção. Compreender as vantagens de cada abordagem, assim como dominar ferramentas como React Native e Flutter, permite ao desenvolvedor tomar decisões mais estratégicas. Experimente baixar um ambiente de desenvolvimento móvel, como o Android Studio ou o Visual Studio Code com plugins específicos para mobile, e teste na prática qual abordagem melhor se adapta aos seus projetos. O universo do desenvolvimento mobile está repleto de oportunidades para inovar e transformar ideias em aplicativos eficientes e impactantes.
      </Text>
    </View>
    <RNGoogleMobileAdsBannerView
      manualImpressionsEnabled={false}
      onNativeEvent={[Function]}
      request="{}"
      sizeConfig={
        {
          "maxHeight": undefined,
          "sizes": [
            "ANCHORED_ADAPTIVE_BANNER",
          ],
          "width": undefined,
        }
      }
      style={
        {
          "height": 0,
          "width": 0,
        }
      }
      unitId="ca-app-pub-3940256099942544/2435281174"
    />
  </View>
</RCTScrollView>
`;
